src/main/java/org/jitsi/videobridge/AbstractEndpoint.java

```diff
@@ -354,8 +386,57 @@ public JSONObject getDebugState()
     public abstract void addRtpExtension(RtpExtension rtpExtension);
 
     /**
-     *
      * @param newVideoConstraints
      */
-    public abstract void setVideoConstraints(Map<String, VideoConstraints> newVideoConstraints);
+    public abstract void setSenderVideoConstraints(ImmutableMap<String, VideoConstraints> newVideoConstraints);
+
+    /**
+     * Notifies this instance that the max video constraints that the bridge
+     * needs to receive from this endpoint has changed. Each implementation
+     * handles this notification differently.
+     *
+     * @param maxVideoConstraints the max video constraints that the bridge
+     *                            needs to receive from this endpoint
+     */
+    protected abstract void
+    onMaxReceiverVideoConstraintsChanged(VideoConstraints maxVideoConstraints);
+
+    /**
+     * The map of receiver endpoint id -> video constraints.
+     */
+    private final Map<String, VideoConstraints> receiverVideoConstraintsMap = new ConcurrentHashMap<>();
+
+    /**
+     * Notifies this instance that the specified endpoint wants to receive
+     * the specified video constraints from the endpoint attached to this
+     * instance (the sender).
+     *
+     * @param endpointId          the id that specifies the receiver endpoint
+     * @param newVideoConstraints the video constraints that the receiver
+     *                            wishes to receive.
+     */
+    public void addReceiver(String endpointId, VideoConstraints newVideoConstraints)
+    {
+        VideoConstraints oldVideoConstraints = receiverVideoConstraintsMap.put(endpointId, newVideoConstraints);
+        if (oldVideoConstraints == null
+            || !oldVideoConstraints.equals(newVideoConstraints))
+        {
+            setReceiverVideoConstraints(ImmutableMap.copyOf(receiverVideoConstraintsMap));
```

Comment:
What is the difference between adding these constraints to `receiverVideoConstraintsMap` and calling `setReceiverVideoConstraints` ?

Reply:
`setReceiverVideoConstraints` (which I renamed to `receiverVideoConstraintsChanged`) is called when the receiver video constraints change (which happens when a receiver is added or removed to/from the map) and it computes the new max receiver video constraints. If that's different from the old (current) max receiver video constraints, then it calls the `maxReceiverVideoConstraints` method, effectively yielding to the implementation class.